### PR TITLE
[FLINK-21223][python] Support to specify the output types of Python UDFs via string

### PR DESCRIPTION
**Pull request by HuangXingBo**
## What is the purpose of the change

*This pull request will support to specify the output types of Python UDFs via string*

## Brief change log

  - *Support to specify the output types of Python UDFs via string*

## Verifying this change

This change added tests and can be verified as follows:

  - *`test_all_data_types_string` and orginal tests in `test_udtf` and `test_udaf`*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by HuangXingBo** [comment]
@flinkbot run azure

File: flink-python/pyflink/table/udf.py [hunk 1]
```diff
@@ -516,8 +539,14 @@ def _create_judf(self, serialized_func, j_input_types, j_function_kind):
             gateway = get_gateway()
             j_input_types = java_utils.to_jarray(
                 gateway.jvm.DataType, [_to_java_data_type(i) for i in self._input_types])
```
Comment (dianfu) [inline flink-python/pyflink/table/udf.py:541]: Will we support specifying input types via string?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/python/PythonAggregateFunction.java [hunk 2]
```diff
@@ -120,6 +163,15 @@ public TypeInference getTypeInference(DataTypeFactory typeFactory) {
         if (inputTypes != null) {
             builder.typedArguments(inputTypes);
         }
+
+        if (resultType == null) {
```
Comment (dianfu) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/python/PythonAggregateFunction.java:167]: What about moving this to the constructor?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/python/PythonTableAggregateFunction.java [hunk 3]
```diff
@@ -120,6 +163,15 @@ public TypeInference getTypeInference(DataTypeFactory typeFactory) {
         if (inputTypes != null) {
             builder.typedArguments(inputTypes);
         }
+
```
Comment (dianfu) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/python/PythonTableAggregateFunction.java:166]: `line 152: TypeConversions.fromDataTypeToLegacyInfo(resultType)`

Does it still work?

**Review by dianfu — APPROVED** [review]

**Pull request closed without merging**